jdk-http-client/src/main/scala/org/http4s/client/jdkhttpclient/JdkHttpClient.scala

```diff
@@ -0,0 +1,131 @@
+import org.http4s.util.CaseInsensitiveString
+import org.http4s.{Header, Headers, HttpVersion, Request, Response, Status}
+import org.reactivestreams.FlowAdapters
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+
+// TODO documentation
+object JdkHttpClient {
+
+  def apply[F[_]](
+      jdkHttpClient: HttpClient,
+      ignoredHeaders: Set[CaseInsensitiveString] = restrictedHeaders
+  )(implicit F: ConcurrentEffect[F]): Client[F] = {
+
+    def convertRequest(req: Request[F]): F[HttpRequest] =
+      convertHttpVersionFromHttp4s[F](req.httpVersion).map { version =>
+        val rb = HttpRequest.newBuilder
+          .method(
+            req.method.name, {
+              val publisher = FlowAdapters.toFlowPublisher(
+                StreamUnicastPublisher(req.body.chunks.map(_.toByteBuffer)))
+              if (req.isChunked)
+                BodyPublishers.fromPublisher(publisher)
+              else
+                req.contentLength.fold(BodyPublishers.noBody)(
+                  BodyPublishers.fromPublisher(publisher, _))
+            }
+          )
+          .uri(URI.create(req.uri.renderString))
+          .version(version)
+        val headers = req.headers.iterator
+        // hacky workaround (see e.g. https://stackoverflow.com/questions/53979173)
+          .filterNot(h => ignoredHeaders.contains(h.name))
+          .flatMap(h => Iterator(h.name.value, h.value))
+          .toArray
+        (if (headers.isEmpty) rb else rb.headers(headers: _*)).build
+      }
+
+    def convertResponse(res: HttpResponse[Flow.Publisher[util.List[ByteBuffer]]]): F[Response[F]] =
+      F.fromEither(Status.fromInt(res.statusCode)).map { status =>
+        Response(
+          status = status,
+          headers = Headers(res.headers.map.asScala.flatMap {
+            case (k, vs) => vs.asScala.map(Header(k, _))
+          }.toList),
+          httpVersion = res.version match {
+            case HttpClient.Version.HTTP_1_1 => HttpVersion.`HTTP/1.1`
+            case HttpClient.Version.HTTP_2 => HttpVersion.`HTTP/2.0`
+          },
+          body = FlowAdapters
+            .toPublisher(res.body)
+            .toStream[F]
+            .flatMap(bs =>
+              Stream.fromIterator(bs.asScala.map(Chunk.byteBuffer).iterator).flatMap(Stream.chunk))
+        )
+      }
+
+    Client[F] { req =>
+      Resource.liftF(
+        convertRequest(req)
+          .flatMap(r =>
+            fromCompletableFuture(F.delay(jdkHttpClient.sendAsync(r, BodyHandlers.ofPublisher))))
+          .flatMap(convertResponse))
+    }
+  }
+
+  def createUnderlyingClient[F[_]](
+      httpVersion: HttpVersion = HttpVersion.`HTTP/1.1`,
+      connectTimeout: Duration = 10.seconds,
+      redirectPolicy: HttpClient.Redirect = HttpClient.Redirect.NEVER,
+      sslContext: => SSLContext = SSLContext.getDefault,
+      proxySelector: => ProxySelector = ProxySelector.getDefault,
+      customEC: Option[Executor] = None
```

Comment:
IMHO the default executor created for the `HttpClient` is a good default. I decided to take an `Executor` because the main way to create an `ExecutionContext` is to use `fromExecutor` or `fromExecutorService`, and either way it is very easy to convert between them.